flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveSchedulerITCase.java

```diff
@@ -231,6 +232,43 @@ public void testStopWithSavepointFailOnFirstSavepointSucceedOnSecond() throws Ex
         assertThat(savepoint, containsString(savepointDirectory.getAbsolutePath()));
     }
 
+    @Test
+    public void testCancellationOfJobInRestartLoop() throws Exception {
+        final long timeInRestartingState = 10000L;
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+        // configure a high delay between attempts: We'll stay in RESTARTING for 10 seconds.
+        env.setRestartStrategy(
+                RestartStrategies.fixedDelayRestart(Integer.MAX_VALUE, timeInRestartingState));
+        env.addSource(new NotifyOnRunningAndFailingSource()).addSink(new DiscardingSink<>());
+
+        JobClient client = env.executeAsync();
+
+        NotifyOnRunningAndFailingSource.runningLatch.await();
+
+        // wait until we are in RESTARTING state
+        CommonTestUtils.waitUntilCondition(
+                () -> client.getJobStatus().get() == JobStatus.RESTARTING,
+                Deadline.fromNow(Duration.of(timeInRestartingState, ChronoUnit.MILLIS)),
+                5);
+
+        // now cancel while in RESTARTING state
+        client.cancel().get();
+    }
```

Comment:
Nit: Does this really have to live in the `flink-tests` module? It feels to me that this should be testable as an integration test in the `flink-runtime` module.